OpenRA.Game/Actor.cs

```diff
@@ -140,43 +141,57 @@ internal Actor(World world, string name, TypeDictionary initDict)
 					throw new NotImplementedException("No rules definition for unit " + name);
 
 				Info = world.Map.Rules.Actors[name];
-				foreach (var trait in Info.TraitsInConstructOrder())
-				{
-					AddTrait(trait.Create(init));
 
-					// Some traits rely on properties provided by IOccupySpace in their initialization,
-					// so we must ready it now, we cannot wait until all traits have finished construction.
-					if (trait is IOccupySpaceInfo)
-						OccupiesSpace = Trait<IOccupySpace>();
+				var resolveOrdersList = new List<IResolveOrder>();
+				var renderModifiersList = new List<IRenderModifier>();
+				var rendersList = new List<IRender>();
+				var mouseBoundsList = new List<IMouseBounds>();
+				var visibilityModifiersList = new List<IVisibilityModifier>();
+				var becomingIdlesList = new List<INotifyBecomingIdle>();
+				var tickIdlesList = new List<INotifyIdle>();
+				var targetablesList = new List<ITargetable>();
+				var targetablePositionsList = new List<ITargetablePositions>();
+				var synchHashesList = new List<SyncHash>();
+
+				foreach (var traitInfo in Info.TraitsInConstructOrder())
+				{
+					var trait = traitInfo.Create(init);
+					AddTrait(trait);
+
+					// PERF: Cache all these traits as soon as the actor is created. This is a fairly cheap one-off cost per
+					// actor that allows us to provide some fast implementations of commonly used methods that are relied on by
+					// performance-sensitive parts of the core game engine, such as pathfinding, visibility and rendering.
+					// Note: The blocks are required to limit the scope of the t's, so we make an exception to our normal style
+					// rules for spacing in order to keep these assignments compact and readable.
+					{ if (trait is IPositionable t) positionable = t; }
+					{ if (trait is IOccupySpace t) OccupiesSpace = t; }
+					{ if (trait is IEffectiveOwner t) EffectiveOwner = t; }
+					{ if (trait is IFacing t) facing = t; }
+					{ if (trait is IHealth t) health = t; }
+					{ if (trait is IResolveOrder t) resolveOrdersList.Add(t); }
+					{ if (trait is IRenderModifier t) renderModifiersList.Add(t); }
+					{ if (trait is IRender t) rendersList.Add(t); }
+					{ if (trait is IMouseBounds t) mouseBoundsList.Add(t); }
+					{ if (trait is IVisibilityModifier t) visibilityModifiersList.Add(t); }
+					{ if (trait is IDefaultVisibility t) defaultVisibility = t; }
+					{ if (trait is INotifyBecomingIdle t) becomingIdlesList.Add(t); }
+					{ if (trait is INotifyIdle t) tickIdlesList.Add(t); }
+					{ if (trait is ITargetable t) targetablesList.Add(t); }
+					{ if (trait is ITargetablePositions t) targetablePositionsList.Add(t); }
+					{ if (trait is ISync t) synchHashesList.Add(new SyncHash(t)); }
```

Comment:
Possible cleaner with a switch.
```
 switch (trait)
    {
        case IPositionable t:
           positionable = t;
           break;
        case IOccupySpace t:
           OccupiesSpace = t; 
           break;
.
.
.
```

Reply:
That was my first idea too; it doesn't work because many traits implement more than one interface!